### PR TITLE
chore(metadata): removes access_id from metadata

### LINKED ISSUE #9050: IDEA: Ignore metadata access in 3.0
Basically the API wouldn't change but md.access_id would be ignored on queries and in the metadata cache. Always reads a public. Baby step to removing it from the API completely.
- Deprecation PR 2.x: #9069
- Implementation PR master: #10270

### PR DESCRIPTION
**Pull request by mrclay**
Elgg now ignores the access_id fields of metadata, hence all metadata set previously are now visible to queries in all contexts, including logged out.

Methods to set metadata no longer accept `$access_id` arguments, or values given are ignored. ElggMetadata objects always return ACCESS_PUBLIC when the "access_id" property is read.

User profile field data is now stored in annotations, but a copy is maintained in metadata for better BC with older plugins. A field with the name "example" will be stored in annotation(s) with name "profile:example". At the time of upgrade, active profile fields will be migrated to annotation storage, but inactive profile field data will be left as inert metadata only.

Fixes #9050

BREAKING CHANGE:
Plugins can no longer rely on Elgg to "hide" metadata in queries. All metadata is assumed to be public. Plugins that read user profile fields in metadata will see all fields every time, and plugins that write user profile fields in metadata will have no effect. These plugins should instead access fields via annotations; see the profile edit actions and forms for reference.

- [x] get #10879 in
- [x] async upgrade that requires the "profile" plugin be enabled and any fields you want migrated must be active.
- [x] Steve tested upgrade

### CODE REVIEW

**Comment by mrclay** [comment]
Are we worried about name collisions with annotations starting with `profile:*`?

**Review by hypeJunction — APPROVED** [review]

**Comment by hypeJunction** [comment]
Likely not

**Comment by mrclay** [comment]
@hypeJunction I'm having trouble debugging metadata mocks here

**Comment by hypeJunction** [comment]
Checking now

**Comment by mrclay** [comment]
phpunit suite on my machine runs out of memory, which is a bummer

**Comment by hypeJunction** [comment]
Found the problem, there are new brackets in the metadata query SQL, so the query specs were not matching... Will make a PR

**Comment by mrclay** [comment]
I rebased, sorry.

**Comment by hypeJunction** [comment]
Woohoo

File: mod/search/search_hooks.php
Comment (jdalsem) [inline mod/search/search_hooks.php]: can't we do this a bit smarter... we can get all annotations at once right?

File: views/default/forms/profile/edit.php
Comment (jdalsem) [inline views/default/forms/profile/edit.php]: similar as previous comment

**Review by jdalsem — APPROVED** [review]

File: engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php
Comment (jdalsem) [inline engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php]: just wondering... what if the profile plugin is not enabled at this moment??? the hook for profile fields is not triggered and that will cause some issues when you activate profile again...

File: views/default/forms/profile/edit.php
Comment (mrclay) [inline views/default/forms/profile/edit.php]: Probably not worth optimizing an edit form. Goal was to keep rewrite easy to review, too.

File: engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php
Comment (mrclay) [inline engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php]: True. Also fields that aren't currently present aren't migrated. Got any solutions to either?

File: views/default/forms/profile/edit.php
Comment (jdalsem) [inline views/default/forms/profile/edit.php]: lets make an extra ticket for it then... as it is 1 query each profile field... some forms have a lot of them

Comment (jdalsem) [inline views/default/forms/profile/edit.php]: we need to set the best example... otherwise people will keep copying bad habits

File: engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php
Comment (jdalsem) [inline engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php]: can we require profile for the upgrade?? maybe move it to a manual upgrade process to be able to do so? That way you can prepare for the profile field migration....

Comment (mrclay) [inline engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php]: We could move the profile field registration to core, or just duplicate the code here. Or rewrite as async migration class and demand profile be enabled. :(

Comment (mrclay) [inline engine/lib/upgrades/2016092301-0.61-profiles_to_annotations-3b68ae209aa07beb.php]: I'm sitting in a doctor's office not getting realtime responses :)

**Comment by jdalsem** [comment]
Could we finish this up? It is almost done...

**Comment by mrclay** [comment]
I guess this is now blocked by #10513. ☹️

**Comment by jdalsem** [comment]
no it is not... just do a traditional upgrade like you already have made... did the same with the site_guid removal

**Comment by jdalsem** [comment]
or maybe it is blocked as we talked about the async job to have all the profile fields available...

**Comment by mrclay** [comment]
OK I'll check your PR.

**Comment by hypeJunction** [comment]
Can we resolve the conflicts and get this in?

**Comment by jdalsem** [comment]
still need to figure out how to get all profile fields correctly migrated (with an async upgrade script)

**Comment by mrclay** [comment]
I can rebase this, but I've already got a ton of time in this. If someone thinks the upgrade should run async, maybe they can contribute that conversion.

**Comment by jdalsem** [comment]
well without a way to migrate all profile fields, this can not go in as you will loose the access of the profile fields, so we need a way to tackle that

**Comment by hypeJunction** [comment]
We can write an upgrade script if you rebase it. It won't be that much work now that we have upgrade API

**Comment by mrclay** [comment]
> well without a way to migrate all profile fields

It does migrate them, just not async. I'll work on this soon.

File: mod/search/search_hooks.php
Comment (mrclay) [inline mod/search/search_hooks.php]: Let's optimize this later.

**Comment by mrclay** [comment]
The profile metadata upgrade is now async and in the profiles plugin. It warns the admin to enable any profiles (s)he wants migrated.

**Comment by mrclay** [comment]
@jdalsem @jeabakker @hypeJunction I've tested the migration locally, and all test are passing. Please give this a go. I'll merge Apr 14 unless big issues are found.

**Comment by mrclay** [comment]
👍  Upgrade from 2.3 automatically queues up core and plugin async upgrades.

**Comment by jdalsem** [comment]
please let met test this before merge...

**Comment by mrclay** [comment]
I've worked on this since [Oct 2015](https://github.com/Elgg/Elgg/issues/9050#issuecomment-150397759), not waiting any longer, sorry.

**Comment by jdalsem** [comment]
main concern was with access_id being lost during migration, but i see you left the column intact. So that should cause no issue migrating profile data with the correct access_id to annotations.

**Pull request merged**
